File: Apps/Android/app/src/main/kotlin/com/opencapture/openzcine/transport/InfrastructureDiscovery.kt
package com.opencapture.openzcine.transport

/**
 * What a single dial at one host actually did.
 *
 * Twin of the shared core's `HostProbeVerdict`. On iOS this was a `String` passed between the
 * transport, two sweeps and the tally, and it drifted — a verdict nobody emitted was counted as
 * occupancy while running out of file descriptors counted as evidence the network was fine. A
 * miscounted verdict is not cosmetic: it is what the operator is told to go and do next.
 */
public enum class HostProbeVerdict {
    /** TCP accepted. Something is serving this port. */
    OPEN,

    /** RST. A host is there and is not serving this port — a router, a laptop, a TV. */
    REFUSED,

    /** No answer inside the budget. Nobody holds the address, or nothing can reach it. */
    TIMEOUT,

    /** No route to the host — on-link and unanswered at the link layer. */
    NO_ROUTE,

    /** No route to that network at all. */
    NO_NETWORK,

    /** The OS refused us the dial; says nothing about the host. */
    DENIED,

    /** Anything else, including running out of descriptors. Never evidence about the host. */
    UNREACHABLE_OTHER,
    ;

    /**
     * Whether SOMETHING holds this address.
     *
     * Only an answer proves occupancy. A refusal is an answer; silence and a denied dial are not.
     */
    public val isOccupied: Boolean
        get() =
            when (this) {
                OPEN, REFUSED -> true
                TIMEOUT, NO_ROUTE, NO_NETWORK, DENIED, UNREACHABLE_OTHER -> false
            }

    /** Whether this verdict is the network declining to carry us, rather than a quiet host. */
    public val isSilence: Boolean
        get() =
            when (this) {
                TIMEOUT, NO_ROUTE, NO_NETWORK -> true
                OPEN, REFUSED, DENIED, UNREACHABLE_OTHER -> false
            }
}

/** Whether an infrastructure (same-LAN Wi‑Fi) search is allowed to run sockets. */
public sealed interface InfrastructurePreflight {
    /** At least one scannable IPv4 interface; search may proceed. */
    public data class Ready(val subnets: List<String>) : InfrastructurePreflight

    /** No private IPv4 interface is up — Wi‑Fi off, mobile data only, or no link. */
    public data object NoScannableInterface : InfrastructurePreflight

    /** The OS is withholding local-network access; LAN sockets will fail or time out. */
    public data object LocalNetworkDenied : InfrastructurePreflight
}

/**
 * Why a completed infrastructure search found no usable camera.
 *
 * Twin of the shared core's `InfrastructureMissReason`. These are the diagnoses that replace an
 * endless "Searching…" — the whole point of classifying a miss rather than just reporting one.
 */
public sealed interface InfrastructureMissReason {
    public data object LocalNetworkDenied : InfrastructureMissReason

    public data object NoScannableInterface : InfrastructureMissReason

    /** Phone is still on the body's own access point — an infrastructure setup can never fulfil. */
    public data object OnCameraAccessPoint : InfrastructureMissReason

    /** Almost no on-link activity — wrong network, or clients kept apart from each other. */
    public data object NetworkUnreachable : InfrastructureMissReason

    /** Hosts answer on the subnet, but nothing accepts PTP-IP. */
    public data object HostsVisibleNoPtp : InfrastructureMissReason

    /** Search budget exhausted without a camera, and without a stronger diagnosis. */
    public data object CameraNotFound : InfrastructureMissReason

    /** Another OpenZCine device holds the only known body. */
    public data class HeldByOtherDevice(val holder: String) : InfrastructureMissReason
}

/**
 * Counts from one occupancy (mute :15740) pass — the inputs classification needs, not a log.
 *
 * Twin of the shared core's `InfrastructureSweepTally`.
 */
public data class InfrastructureSweepTally(
    val openCount: Int = 0,
    val occupiedCount: Int = 0,
    val timeoutCount: Int = 0,
    val refusedCount: Int = 0,
    val noRouteCount: Int = 0,
    val deniedCount: Int = 0,
    val otherCount: Int = 0,
    val hostsProbed: Int = 0,
    val occupiedHosts: List<String> = emptyList(),
    val openHosts: List<String> = emptyList(),
) {
    /**
     * Merge another pass's counts in, keeping the union of what each pass saw.
     *
     * A body waking mid-search answers on pass two and not pass one; classification must see both.
     * This exists instead of hand-patching individual fields at the call site, which is where iOS
     * came to set `hostsProbed` from the OCCUPIED count — so a sweep where nothing answered
     * reported one host probed, and "nothing on this network answered" could never be reached.
     */
    public fun merging(other: InfrastructureSweepTally): InfrastructureSweepTally =
        InfrastructureSweepTally(
            openCount = maxOf(openCount, other.openCount),
            occupiedCount = maxOf(occupiedCount, other.occupiedCount),
            timeoutCount = maxOf(timeoutCount, other.timeoutCount),
            refusedCount = maxOf(refusedCount, other.refusedCount),
            noRouteCount = maxOf(noRouteCount, other.noRouteCount),
            deniedCount = maxOf(deniedCount, other.deniedCount),
            otherCount = maxOf(otherCount, other.otherCount),
            hostsProbed = maxOf(hostsProbed, other.hostsProbed),
            occupiedHosts = (occupiedHosts + other.occupiedHosts).distinct().sorted(),
            openHosts = (openHosts + other.openHosts).distinct().sorted(),
        )

    public companion object {
        /**
         * Build a tally from per-host verdicts. Exhaustive by construction: a verdict added to
         * [HostProbeVerdict] cannot be forgotten here without the compiler saying so.
         */
        public fun from(verdicts: Map<String, HostProbeVerdict>): InfrastructureSweepTally {
            var open = 0
            var refused = 0
            var timeout = 0
            var noRoute = 0
            var denied = 0
            var other = 0
            val openHosts = mutableListOf<String>()
            val occupiedHosts = mutableListOf<String>()
            for ((host, verdict) in verdicts) {
                when (verdict) {
                    HostProbeVerdict.OPEN -> open += 1
                    HostProbeVerdict.REFUSED -> refused += 1
                    HostProbeVerdict.TIMEOUT -> timeout += 1
                    HostProbeVerdict.NO_ROUTE, HostProbeVerdict.NO_NETWORK -> noRoute += 1
                    HostProbeVerdict.DENIED -> denied += 1
                    HostProbeVerdict.UNREACHABLE_OTHER -> other += 1
                }
                if (verdict == HostProbeVerdict.OPEN) openHosts.add(host)
                if (verdict.isOccupied) occupiedHosts.add(host)
            }
            return InfrastructureSweepTally(
                openCount = open,
                occupiedCount = occupiedHosts.size,
                timeoutCount = timeout,
                refusedCount = refused,
                noRouteCount = noRoute,
                deniedCount = denied,
                otherCount = other,
                hostsProbed = verdicts.size,
                occupiedHosts = occupiedHosts.sorted(),
                openHosts = openHosts.sorted(),
            )
        }
    }
}

/** One infrastructure search pass — cameras if any, else a typed miss and the evidence. */
public data class InfrastructureSearchReport(
    val cameras: List<DiscoveredCamera> = emptyList(),
    val miss: InfrastructureMissReason? = null,
    val occupiedHosts: List<String> = emptyList(),
    val openPtpHosts: List<String> = emptyList(),
    val localSubnets: List<String> = emptyList(),
    val directedHosts: List<String> = emptyList(),
) {
    public val foundCamera: Boolean
        get() = cameras.isNotEmpty()
}

/**
 * Pure infrastructure discovery policy — budgets, classification, operator copy.
 *
 * No sockets. The platform finder executes probes; this decides what the outcomes mean and how
 * patient a directed dial should be. Twin of the shared core's `InfrastructureDiscovery`; both are
 * tested against the same table, so a network can never mean one diagnosis on iPhone and another
 * on Android. Change one, change both.
 *
 * The budgets are not guesses. An idle ZR on Wi‑Fi answers a dial in ~1.0–1.2 s, measured on
 * hardware 2026-08-04, because its radio is in power-save and listens only on its beacon cadence.
 * Anything shorter measures the app's patience rather than the camera's presence.
 */
public object InfrastructureDiscovery {
    /** How many times a directed (known) host is dialled before giving up that address. */
    public const val DIRECTED_ATTEMPT_COUNT: Int = 3

    /** Per-attempt connect budget for a directed host (ms). */
    public const val DIRECTED_TIMEOUT_MILLIS: Int = 2_000

    /** Blind mute-scan budget per host (ms) — measured power-save answer time plus margin. */
    public const val BLIND_SWEEP_TIMEOUT_MILLIS: Int = 1_500

    /** Concurrent mute dials in the occupancy sweep. */
    public const val SWEEP_CONCURRENCY: Int = 48

    /** How many full local-/24 occupancy passes before declaring a miss. */
    public const val LOCAL_SWEEP_PASSES: Int = 2

    /**
     * Hosts worth a patient directed probe: candidates filtered to the current local subnets.
     *
     * A sibling setup on another network — a portable travel router left in the saved list — must
     * not be dialled while adding a new Wi‑Fi setup. Dialling it is not a shortcut to the new
     * setup, it is a connect to somewhere else that fails and reads as the new setup failing.
     */
    public fun directedHosts(
        candidates: List<String>,
        localSubnets: List<String>,
        excluded: Set<String> = emptySet(),
    ): List<String> {
        val subnets = localSubnets.toSet()
        val seen = mutableSetOf<String>()
        val output = mutableListOf<String>()
        for (raw in candidates) {
            val host = raw.trim()
            if (!CameraDiscovery.isDialableHost(host)) continue
            if (host in excluded || host in seen) continue
            val base = CameraDiscovery.subnetBase(host) ?: continue
            if (base !in subnets) continue
            seen.add(host)
            output.add(host)
        }
        return output
    }

    /** Classify an empty (or preflight-failed) infrastructure search. */
    public fun classifyMiss(
        preflight: InfrastructurePreflight,
        onCameraAccessPoint: Boolean,
        tally: InfrastructureSweepTally?,
        heldHolderName: String? = null,
    ): InfrastructureMissReason {
        if (preflight is InfrastructurePreflight.LocalNetworkDenied) {
            return InfrastructureMissReason.LocalNetworkDenied
        }
        if (preflight is InfrastructurePreflight.NoScannableInterface) {
            return InfrastructureMissReason.NoScannableInterface
        }
        if (onCameraAccessPoint) return InfrastructureMissReason.OnCameraAccessPoint
        if (!heldHolderName.isNullOrBlank()) {
            return InfrastructureMissReason.HeldByOtherDevice(heldHolderName)
        }
        if (tally == null || tally.hostsProbed <= 0) {
            return InfrastructureMissReason.CameraNotFound
        }
        if (tally.deniedCount > 0 && tally.occupiedCount == 0 && tally.openCount == 0) {
            return InfrastructureMissReason.LocalNetworkDenied
        }
        if (tally.occupiedCount == 0 && tally.openCount == 0) {
            // All timeout / no-route — cannot prove the LAN is live.
            val silent = tally.timeoutCount + tally.noRouteCount
            return if (silent >= maxOf(1, tally.hostsProbed / 2)) {
                InfrastructureMissReason.NetworkUnreachable
            } else {
                InfrastructureMissReason.CameraNotFound
            }
        }
        // The LAN is live (hosts refuse or answer) but nothing speaks PTP-IP.
        if (tally.openCount == 0 && tally.occupiedCount > 0) {
            return InfrastructureMissReason.HostsVisibleNoPtp
        }
        return InfrastructureMissReason.CameraNotFound
    }

    /**
     * Operator-facing sentence for a miss.
     *
     * Short, and in the operator's words. The first iOS cut named the subnet in dotted notation,
     * spelled the camera's menus out as arrow chains and explained port 15740 — all true, and none
     * of it what somebody standing in front of a camera needs. One thing to go and do, and the
     * address stays in the log where it is actually useful.
     *
     * Only [InfrastructureMissReason.LocalNetworkDenied] reads differently from iOS, because the
     * permission it names is a different one: iOS has a Local Network switch, Android gates nearby
     * Wi‑Fi devices instead.
     */
    public fun operatorCopy(
        reason: InfrastructureMissReason,
        cameraName: String? = null,
    ): String {
        val body = cameraName?.takeIf(String::isNotBlank) ?: "the camera"
        return when (reason) {
            is InfrastructureMissReason.LocalNetworkDenied ->
                "OpenZCine needs permission to find devices on your Wi‑Fi. Check its permissions in Settings."
            is InfrastructureMissReason.NoScannableInterface ->
                "This phone isn't on Wi‑Fi. Join the same network as $body."
            is InfrastructureMissReason.OnCameraAccessPoint ->
                "You're on $body's own Wi‑Fi. Join the network you want this setup to use, and we'll keep looking."
            is InfrastructureMissReason.NetworkUnreachable ->
                // Names the cause that actually bit, rather than a generic "check your Wi-Fi". A
                // Wi-Fi 7 router running 6 GHz with MLO put a phone and a camera on links it would
                // not bridge: same network, same subnet, addresses from the same DHCP server, and
                // no packet between them. Days went into that one, with the app saying
                // "Searching…" throughout. Naming a band an operator can switch off is worth it.
                "Nothing on this network answered. Check both are on the same Wi‑Fi — a guest network or a 6 GHz band can keep them apart."
            is InfrastructureMissReason.HostsVisibleNoPtp ->
                "Other devices answer here, but $body doesn't. On the camera, start Connect to computer."
            is InfrastructureMissReason.CameraNotFound ->
                "Looking for $body on this Wi‑Fi. On the camera, start Connect to computer."
            is InfrastructureMissReason.HeldByOtherDevice ->
                "$body is in use by ${reason.holder}."
        }
    }
}

File: Sources/OpenZCineCore/InfrastructureDiscovery.swift
import Foundation

// MARK: - Preflight

/// Whether an infrastructure (same-LAN Wi‑Fi) search is allowed to run sockets.
public enum InfrastructurePreflight: Equatable, Sendable {
    /// At least one scannable IPv4 interface; search may proceed.
    case ready(interfaces: [LocalIPv4Interface])
    /// No private IPv4 interface is up — Wi‑Fi off, cellular only, or no link.
    case noScannableInterface
    /// iOS Local Network permission is denied; LAN sockets will time out or fail.
    case localNetworkDenied
}

// MARK: - Miss reasons

/// Why a completed infrastructure search found no usable camera.
///
/// Ordered for classification priority (most actionable / earliest exit first). These are the
/// operator-facing diagnoses industry LAN tools surface instead of an endless "Searching…".
public enum InfrastructureMissReason: Equatable, Sendable {
    case localNetworkDenied
    case noScannableInterface
    /// Phone is still on the body's own access point — infrastructure shape can never fulfill.
    case onCameraAccessPoint
    /// Almost no on-link activity (gateway and hosts time out) — wrong network, isolation, or
    /// permission that presents as total silence.
    case networkUnreachable
    /// Hosts answer ARP/TCP on the subnet, but nothing accepts PTP-IP (:15740).
    /// Camera not in Connect to PC, or client isolation between phone and body.
    case hostsVisibleNoPTP
    /// Search budget exhausted without a camera (and without a stronger diagnosis).
    case cameraNotFound
    /// Another OpenZCine device holds the only known body.
    case heldByOtherDevice(holder: String)
}

// MARK: - Host probe verdict

/// What a single dial at one host actually did.
///
/// This was a `String` passed between the transport, two sweeps and the tally, and it had already
/// drifted: the tally handled a `"reset"` nobody produced, while `"no-fds"` and `"errno-…"` fell
/// into `other` and quietly counted as evidence the network was fine. A miscounted verdict is not
/// a cosmetic problem — it is what the operator is told to go and do next.
public enum HostProbeVerdict: String, Codable, Sendable, CaseIterable {
    /// TCP accepted. Something is serving this port.
    case open
    /// RST. A host is there and is not serving this port — a router, a laptop, a TV.
    case refused
    /// No answer inside the budget. Nobody holds the address, or nothing can reach it.
    case timeout
    /// EHOSTUNREACH — on-link and unanswered at the link layer.
    case noRoute
    /// ENETUNREACH — no route to that network at all.
    case noNetwork
    /// EACCES / EPERM. The OS refused us the dial; says nothing about the host.
    case denied
    /// Anything else, including running out of descriptors. Never evidence about the host.
    case unreachableOther

    /// Whether SOMETHING holds this address.
    ///
    /// Only an answer proves occupancy. A refusal is an answer; silence and a denied dial are not.
    public var isOccupied: Bool {
        switch self {
        case .open, .refused: return true
        case .timeout, .noRoute, .noNetwork, .denied, .unreachableOther: return false
        }
    }

    /// Whether this verdict is the network declining to carry us, rather than a quiet host.
    public var isSilence: Bool {
        switch self {
        case .timeout, .noRoute, .noNetwork: return true
        case .open, .refused, .denied, .unreachableOther: return false
        }
    }
}

// MARK: - Sweep tally (pure input to classification)

/// Counts from one occupancy (mute :15740) pass — the inputs classification needs, not a log.
public struct InfrastructureSweepTally: Equatable, Sendable {
    public var openCount: Int
    public var occupiedCount: Int
    public var timeoutCount: Int
    public var refusedCount: Int
    public var noRouteCount: Int
    public var deniedCount: Int
    public var otherCount: Int
    public var hostsProbed: Int
    public var occupiedHosts: [String]
    public var openHosts: [String]

    public init(
        openCount: Int = 0,
        occupiedCount: Int = 0,
        timeoutCount: Int = 0,
        refusedCount: Int = 0,
        noRouteCount: Int = 0,
        deniedCount: Int = 0,
        otherCount: Int = 0,
        hostsProbed: Int = 0,
        occupiedHosts: [String] = [],
        openHosts: [String] = []
    ) {
        self.openCount = openCount
        self.occupiedCount = occupiedCount
        self.timeoutCount = timeoutCount
        self.refusedCount = refusedCount
        self.noRouteCount = noRouteCount
        self.deniedCount = deniedCount
        self.otherCount = otherCount
        self.hostsProbed = hostsProbed
        self.occupiedHosts = occupiedHosts
        self.openHosts = openHosts
    }

    /// Build a tally from per-host verdicts. Exhaustive by construction: a verdict added to
    /// ``HostProbeVerdict`` cannot be forgotten here without the compiler saying so.
    public static func from(verdicts: [String: HostProbeVerdict]) -> InfrastructureSweepTally {
        var tally = InfrastructureSweepTally(hostsProbed: verdicts.count)
        var occupied: [String] = []
        var open: [String] = []
        for (host, verdict) in verdicts {
            switch verdict {
            case .open: tally.openCount += 1
            case .refused: tally.refusedCount += 1
            case .timeout: tally.timeoutCount += 1
            case .noRoute, .noNetwork: tally.noRouteCount += 1
            case .denied: tally.deniedCount += 1
            case .unreachableOther: tally.otherCount += 1
            }
            if verdict == .open { open.append(host) }
            if verdict.isOccupied { occupied.append(host) }
        }
        tally.occupiedCount = occupied.count
        tally.occupiedHosts = occupied.sorted()
        tally.openHosts = open.sorted()
        return tally
    }

    /// Merge another pass's counts in, keeping the union of what each pass saw.
    ///
    /// A body waking mid-search answers on pass two and not pass one; classification must see
    /// both. This replaced hand-patching individual fields at the call site, which is where
    /// `hostsProbed` came to be set from the OCCUPIED count — so a sweep where nothing answered
    /// reported one host probed, and "nothing on this network answered" could never be reached.
    public func merging(_ other: InfrastructureSweepTally) -> InfrastructureSweepTally {
        InfrastructureSweepTally(
            openCount: max(openCount, other.openCount),
            occupiedCount: max(occupiedCount, other.occupiedCount),
            timeoutCount: max(timeoutCount, other.timeoutCount),
            refusedCount: max(refusedCount, other.refusedCount),
            noRouteCount: max(noRouteCount, other.noRouteCount),
            deniedCount: max(deniedCount, other.deniedCount),
            otherCount: max(otherCount, other.otherCount),
            hostsProbed: max(hostsProbed, other.hostsProbed),
            occupiedHosts: Array(Set(occupiedHosts).union(other.occupiedHosts)).sorted(),
            openHosts: Array(Set(openHosts).union(other.openHosts)).sorted()
        )
    }
}

// MARK: - Report

/// One infrastructure search pass — cameras if any, else a typed miss and the evidence.
public struct InfrastructureSearchReport: Equatable, Sendable {
    public var cameras: [DiscoveredCamera]
    public var miss: InfrastructureMissReason?
    public var occupiedHosts: [String]
    public var openPTPHosts: [String]
    public var localSubnets: [String]
    public var directedHosts: [String]

    public init(
        cameras: [DiscoveredCamera] = [],
        miss: InfrastructureMissReason? = nil,
        occupiedHosts: [String] = [],
        openPTPHosts: [String] = [],
        localSubnets: [String] = [],
        directedHosts: [String] = []
    ) {
        self.cameras = cameras
        self.miss = miss
        self.occupiedHosts = occupiedHosts
        self.openPTPHosts = openPTPHosts
        self.localSubnets = localSubnets
        self.directedHosts = directedHosts
    }

    public var foundCamera: Bool { !cameras.isEmpty }
}

// MARK: - Policy

/// Pure infrastructure discovery policy — budgets, classification, operator copy.
///
/// No sockets. Platform finders execute probes; this module decides what the outcomes mean and
/// how patient a directed dial should be (ZR Wi‑Fi power-save is ~1.0–1.2 s per HW 2026-08-04).
public enum InfrastructureDiscovery: Sendable {
    /// How many times a directed (known) host is dialled before giving up that address.
    public static let directedAttemptCount = 3
    /// Per-attempt connect budget for a directed host (ms). Longer than the blind sweep so a
    /// body that woke on the first SYN still has room if the mesh is slow.
    public static let directedTimeoutMilliseconds: UInt64 = 2_000
    /// Blind mute-scan budget per host (ms) — matches measured power-save answer time + margin.
    public static let blindSweepTimeoutMilliseconds: UInt64 = 1_500
    /// Concurrent mute dials in the occupancy sweep.
    public static let sweepConcurrency = 48
    /// Init / name probe after a port opens (ms).
    public static let identifyTimeoutMilliseconds: UInt64 = 2_000
    /// How many full local-/24 occupancy passes before declaring a miss (body may wake mid-search).
    public static let localSweepPasses = 2

    /// Directed patient-probe schedule: `attempts` × `timeoutMilliseconds`.
    public static func patientProbeSchedule(
        attempts: Int = directedAttemptCount,
        timeoutMilliseconds: UInt64 = directedTimeoutMilliseconds
    ) -> (attempts: Int, timeoutMilliseconds: UInt64) {
        (max(1, attempts), timeoutMilliseconds)
    }

    /// Hosts worth a patient directed probe: priority list filtered to current local subnets.
    ///
    /// A sibling setup on another network (e.g. portable travel router) must not be dialled when
    /// adding a new infrastructure setup — that was a field false-fail (`191f3fd6`).
    public static func directedHosts(
        candidates: [String],
        localSubnets: [String],
        excluded: Set<String> = []
    ) -> [String] {
        let subnetSet = Set(localSubnets)
        var seen: Set<String> = []
        var output: [String] = []
        for raw in candidates {
            guard let host = PTPIPPairedHosts.normalizedHost(raw),
                !DiscoveredCamera.isUSBHostKey(host),
                !excluded.contains(host),
                !seen.contains(host),
                let base = CameraDiscovery.subnetBase(for: host),
                subnetSet.contains(base)
            else { continue }
            seen.insert(host)
            output.append(host)
        }
        return output
    }

    /// Classify an empty (or preflight-failed) infrastructure search.
    public static func classifyMiss(
        preflight: InfrastructurePreflight,
        onCameraAccessPoint: Bool,
        tally: InfrastructureSweepTally?,
        heldHolderName: String? = nil
    ) -> InfrastructureMissReason {
        if case .localNetworkDenied = preflight { return .localNetworkDenied }
        if case .noScannableInterface = preflight { return .noScannableInterface }
        if onCameraAccessPoint { return .onCameraAccessPoint }
        if let holder = heldHolderName, !holder.isEmpty {
            return .heldByOtherDevice(holder: holder)
        }
        guard let tally, tally.hostsProbed > 0 else {
            return .cameraNotFound
        }
        // Permission / total blackhole: nothing reachable, many timeouts, or explicit denied.
        if tally.deniedCount > 0, tally.occupiedCount == 0, tally.openCount == 0 {
            return .localNetworkDenied
        }
        if tally.occupiedCount == 0, tally.openCount == 0 {
            // All timeout / no-route — cannot prove the LAN is live.
            let silent = tally.timeoutCount + tally.noRouteCount
            if silent >= max(1, tally.hostsProbed / 2) {
                return .networkUnreachable
            }
            return .cameraNotFound
        }
        // LAN is live (hosts refuse or answer) but nothing speaks PTP-IP.
        if tally.openCount == 0, tally.occupiedCount > 0 {
            return .hostsVisibleNoPTP
        }
        // Ports opened but Init did not yield a camera (unlikely; still a miss).
        return .cameraNotFound
    }

    /// Operator-facing sentence for a miss.
    ///
    /// Short, and in the operator's words. The first cut named the subnet in dotted notation,
    /// spelled the camera's menus out as arrow chains and explained port 15740 — all true, and
    /// none of it what somebody standing in front of a camera needs. One thing to go and do,
    /// and the address stays in the log where it is actually useful.
    public static func operatorCopy(
        for reason: InfrastructureMissReason,
        cameraName: String? = nil
    ) -> String {
        let body = cameraName ?? "the camera"
        switch reason {
        case .localNetworkDenied:
            return "OpenZCine needs Local Network access to find cameras. Turn it on in Settings."
        case .noScannableInterface:
            return "This phone isn't on Wi‑Fi. Join the same network as \(body)."
        case .onCameraAccessPoint:
            return
                "You're on \(body)'s own Wi‑Fi. Join the network you want this setup to use, and we'll keep looking."
        case .networkUnreachable:
            // Names the cause that actually bit, rather than a generic "check your Wi-Fi". A
            // Wi-Fi 7 router running 6 GHz with MLO put this phone and the camera on links it
            // would not bridge: same network, same subnet, addresses from the same DHCP server,
            // and no packet between them. Days went into that one, and the app said "Searching…"
            // throughout. Naming a band an operator can switch off is worth the extra clause.
            return
                "Nothing on this network answered. Check both are on the same Wi‑Fi — a guest network or a 6 GHz band can keep them apart."
        case .hostsVisibleNoPTP:
            return
                "Other devices answer here, but \(body) doesn't. On the camera, start Connect to computer."
        case .cameraNotFound:
            return "Looking for \(body) on this Wi‑Fi. On the camera, start Connect to computer."
        case .heldByOtherDevice(let holder):
            return "\(body) is in use by \(holder)."
        }
    }
}
